File: src/main/java/org/c3lang/intellij/stubs/C3ModuleStub.java
package org.c3lang.intellij.stubs;

import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.IStubElementType;
import com.intellij.psi.stubs.StubBase;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import org.c3lang.intellij.psi.C3Module;
import org.c3lang.intellij.psi.ModuleName;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;

public class C3ModuleStub extends StubBase<C3Module>
{
	private final @Nullable ModuleName module;

	public C3ModuleStub(
		@Nullable StubElement<?> parent,
		@Nullable IStubElementType<?, ?> elementType,
		@Nullable ModuleName module)
	{
		super(parent, elementType);
		this.module = module;
	}

	public C3ModuleStub(
		@NotNull StubElement<? extends PsiElement> parent,
		@NotNull C3ModuleElementType elementType,
		@NotNull C3Module psi)
	{
		this(parent, elementType, ModuleName.from(psi));
	}

	public C3ModuleStub(
		@NotNull StubElement<?> parent,
		@NotNull C3ModuleElementType elementType,
		@NotNull StubInputStream dataStream) throws IOException
	{
		this(parent, elementType, StubStreamExtensions.readModuleName(dataStream));
	}

	public @Nullable ModuleName getModule()
	{
		return module;
	}

	public void serialize(@NotNull StubOutputStream dataStream) throws IOException
	{
		StubStreamExtensions.writeNullableUTFFast(dataStream, module != null ? module.getValue() : null);
	}
}

File: src/main/java/org/c3lang/intellij/stubs/C3ModuleStub.kt
package org.c3lang.intellij.stubs

import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.*
import org.c3lang.intellij.psi.C3Module
import org.c3lang.intellij.psi.ModuleName

class C3ModuleStub(
    parent: StubElement<*>?,
    elementType: IStubElementType<*, *>?,
    val module: ModuleName?,
) : StubBase<C3Module?>(parent, elementType) {

    constructor(
        parent: StubElement<out PsiElement?>,
        elementType: C3ModuleElementType,
        psi: C3Module
    ) : this(
        parent = parent,
        elementType = elementType,
        module = ModuleName.from(psi)
    )

    constructor(
        parent: StubElement<*>,
        elementType: C3ModuleElementType,
        dataStream: StubInputStream
    ) : this(
        parent = parent,
        elementType = elementType,
        module = dataStream.readNullableUTFFast()?.let(::ModuleName)
    )

    fun serialize(dataStream: StubOutputStream) {
        dataStream.writeNullableUTFFast(module?.value)
    }
}
